File: backend/app/schemas/workspace.py
"""
Pydantic schemas for Workspace endpoints.

Supports full canvas save/load: the frontend sends the entire
nodes + edges array on PUT, and the backend replaces atomically.
"""

from datetime import datetime
from typing import Any, Optional, List

from pydantic import BaseModel, Field


# ── Workspace Node ────────────────────────────────────────


class WorkspaceNodeCreate(BaseModel):
    node_key: str = Field(..., max_length=80, description="Stable React Flow node id")
    node_type: str = Field(
        ...,
        description="client | device | fl-server | attack | traffic-source | rate-filter | monitor",
    )
    position_x: float = Field(default=0.0)
    position_y: float = Field(default=0.0)
    data: dict[str, Any] = Field(default_factory=dict, description="Type-specific node data blob")


class WorkspaceNodeOut(BaseModel):
    id: int
    workspace_id: int
    node_key: str
    node_type: str
    position_x: float
    position_y: float
    data: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Workspace Edge ────────────────────────────────────────


class WorkspaceEdgeCreate(BaseModel):
    edge_key: str = Field(..., max_length=80, description="Stable React Flow edge id")
    edge_type: str = Field(
        ...,
        description="ownership | fl-communication | traffic-feed | attack-vector | observation",
    )
    source_key: str = Field(..., max_length=80)
    target_key: str = Field(..., max_length=80)
    data: dict[str, Any] = Field(default_factory=dict)


class WorkspaceEdgeOut(BaseModel):
    id: int
    workspace_id: int
    edge_key: str
    edge_type: str
    source_key: str
    target_key: str
    data: dict[str, Any]
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Workspace ─────────────────────────────────────────────


class ViewportState(BaseModel):
    x: float = 0.0
    y: float = 0.0
    zoom: float = 1.0


class WorkspaceCreate(BaseModel):
    name: str = Field(default="My Workspace", max_length=100)
    description: Optional[str] = None
    nodes: List[WorkspaceNodeCreate] = Field(default_factory=list)
    edges: List[WorkspaceEdgeCreate] = Field(default_factory=list)
    viewport: ViewportState = Field(default_factory=ViewportState)


class WorkspaceSave(BaseModel):
    """
    Full canvas save — atomically replaces name, nodes, edges, and viewport.
    All fields are optional so the frontend can do partial saves.
    """
    name: Optional[str] = Field(default=None, max_length=100)
    description: Optional[str] = None
    nodes: Optional[List[WorkspaceNodeCreate]] = None
    edges: Optional[List[WorkspaceEdgeCreate]] = None
    viewport: Optional[ViewportState] = None


class WorkspaceOut(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    viewport: ViewportState
    created_at: datetime
    updated_at: datetime
    nodes: List[WorkspaceNodeOut] = []
    edges: List[WorkspaceEdgeOut] = []

    model_config = {"from_attributes": True}


class WorkspaceBrief(BaseModel):
    """Lightweight listing — no nodes or edges."""
    id: int
    name: str
    description: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/workspace.ts
/**
 * Workspace API — CRUD for canvas workspaces.
 *
 * Endpoints map to: /api/v1/workspaces/...
 */

import api from './client';

// ── Response types (match backend WorkspaceOut/WorkspaceBrief) ──

export interface WorkspaceNodeOut {
  id: number;
  workspace_id: number;
  node_key: string;
  node_type: string;
  position_x: number;
  position_y: number;
  data: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceEdgeOut {
  id: number;
  workspace_id: number;
  edge_key: string;
  edge_type: string;
  source_key: string;
  target_key: string;
  data: Record<string, unknown>;
  created_at: string;
}

export interface ViewportState {
  x: number;
  y: number;
  zoom: number;
}

export interface WorkspaceOut {
  id: number;
  name: string;
  description: string | null;
  viewport: ViewportState;
  created_at: string;
  updated_at: string;
  nodes: WorkspaceNodeOut[];
  edges: WorkspaceEdgeOut[];
}

export interface WorkspaceBrief {
  id: number;
  name: string;
  description: string | null;
  created_at: string;
  updated_at: string;
}

// ── Request types ──

export interface WorkspaceNodeCreate {
  node_key: string;
  node_type: string;
  position_x: number;
  position_y: number;
  data: Record<string, unknown>;
}

export interface WorkspaceEdgeCreate {
  edge_key: string;
  edge_type: string;
  source_key: string;
  target_key: string;
  data: Record<string, unknown>;
}

export interface WorkspaceCreatePayload {
  name?: string;
  description?: string;
  nodes?: WorkspaceNodeCreate[];
  edges?: WorkspaceEdgeCreate[];
  viewport?: ViewportState;
}

export interface WorkspaceSavePayload {
  name?: string;
  description?: string;
  nodes?: WorkspaceNodeCreate[];
  edges?: WorkspaceEdgeCreate[];
  viewport?: ViewportState;
}

// ── API functions ──

export const workspaceApi = {
  /** List all workspaces (lightweight — no nodes/edges). */
  list: () =>
    api.get<WorkspaceBrief[]>('/workspaces/').then((r) => r.data),

  /** Create a new workspace. */
  create: (data: WorkspaceCreatePayload = {}) =>
    api.post<WorkspaceOut>('/workspaces/', data).then((r) => r.data),

  /** Load a workspace (includes nodes + edges + viewport). */
  get: (id: number) =>
    api.get<WorkspaceOut>(`/workspaces/${id}`).then((r) => r.data),

  /** Full canvas save — atomically replaces nodes, edges, viewport. */
  save: (id: number, data: WorkspaceSavePayload) =>
    api.put<WorkspaceOut>(`/workspaces/${id}`, data).then((r) => r.data),

  /** Delete a workspace. */
  delete: (id: number) =>
    api.delete(`/workspaces/${id}`).then((r) => r.data),
};
